src/java.rs

```diff
@@ -0,0 +1,224 @@
+        'J' => Some("long"),
+        'F' => Some("float"),
+        'D' => Some("double"),
+        'V' => Some("void"),
+        _ => None,
+    }
+}
+
+fn byte_code_type_to_java_type(byte_code_type: &str, mapper: &ProguardMapper) -> Option<String> {
+    let mut chrs = byte_code_type.chars();
+    let token = chrs.next()?;
+    if token == 'L' {
+        // invalid signature
+        let l = chrs.clone().last();
+        if l.is_none() || l.unwrap() != ';' {
+            return None;
+        }
+        chrs.next_back(); // remove final `;`
+        let obfuscated = chrs.collect::<String>().replace('/', ".");
+
+        if let Some(mapped) = mapper.remap_class(&obfuscated) {
+            return Some(mapped.to_string());
+        }
+
+        return Some(obfuscated);
+    } else if token == '[' {
+        let type_sig = chrs.clone().collect::<String>();
+        if !type_sig.is_empty() {
+            return Some(format!(
+                "{}[]",
+                byte_code_type_to_java_type(chrs.collect::<String>().as_str(), mapper)
+                    .unwrap_or_default()
+            ));
+        }
+    } else if let Some(ty) = java_base_types(token) {
+        return Some(ty.to_string());
+    }
+    Some(byte_code_type.to_string())
+}
+
+// parse_obfuscated_bytecode_signature will parse an obfuscated signatures into parameter
+// and return types that can be then deobfuscated
+fn parse_obfuscated_bytecode_signature(signature: &str) -> Option<(Vec<String>, String)> {
+    let mut chrs = signature.chars();
+
+    let token = chrs.next();
+    if token.unwrap_or_default() != '(' {
+        return None;
+    }
+
+    let sig = chrs.collect::<String>();
+    let split_sign = sig.rsplitn(2, ')').collect::<Vec<&str>>();
+    if split_sign.len() != 2 {
+        return None;
+    }
+
+    let return_type = split_sign[0];
+    let parameter_types = split_sign[1];
+    if return_type.is_empty() {
+        return None;
+    }
+
+    let mut types: Vec<String> = Vec::new();
+    let mut tmp_buf: Vec<char> = Vec::new();
+
+    let mut param_chrs = parameter_types.chars();
+    while let Some(token) = param_chrs.next() {
+        if token == 'L' {
+            tmp_buf.push(token);
+            for c in param_chrs.by_ref() {
+                tmp_buf.push(c);
+                if c == ';' {
+                    break;
+                }
+            }
+            if tmp_buf.is_empty() || !tmp_buf.ends_with(&[';']) {
+                return None;
+            }
+            types.push(tmp_buf.iter().collect());
+            tmp_buf.clear();
+        } else if token == '[' {
+            tmp_buf.push('[');
+        } else if let Some(ty) = java_base_types(token) {
+            if !tmp_buf.is_empty() {
+                tmp_buf.append(&mut ty.chars().collect::<Vec<char>>());
+                types.push(tmp_buf.iter().collect());
+                tmp_buf.clear();
+            } else {
+                types.push(token.to_string());
+            }
```

Comment:
```suggestion
            if !tmp_buf.is_empty() {
                tmp_buf.push_str(ty);
                types.push(&tmp_buf);
                tmp_buf.clear();
            } else {
                types.push(token.to_string());
            }
```

I believe you need to use `tmp_buf` because it can have a `[` marker in it?